### ISSUE TITLE
Cucumber tests

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#2545]
#2390 introduces tests using cucumber, and I think we should discuss this aspect of the PR separately.

[Cucumber](https://cucumber.io/) is a syntax for writing tests, as an alternative to the MiniTest syntax (that we use) or RSpec (which is very common in the Ruby community). It aims to be a "natural language" way of writing tests, rather than the code-centric approach of the other formats. Here's an example from #2390 , to give a flavour of the syntax:

```
 Scenario: Edit a microcosm
    Given there is a user "abe@example.com" with name "Abe"
    And this user is an organizer of this microcosm
    When user "abe@example.com" logs in
    And I am on the microcosms page
    And I click the link to "/microcosms/mappingdc/edit"
    And I set the microcosm in ".edit_microcosm" to "Baltimore", "40", "-76"
    And I submit the form
    Then I should not see "Washington, DC, USA"
    Then I should see "Baltimore"
```

It's often written for "system" (or "feature") level tests, that is, tests that use a web browser to read the page output, push buttons and fill in forms, and make assertions based on the results. I tend not to find it used much for the other types of test, like "model", "mailer", "controller" tests and so on.

Cucumber claims that its syntax has a number of advantages, and that being written in something approaching natural language makes it easier for non-technical people to read and contribute to the tests. It's often aimed at "Product Owners" and similar positions in larger companies.

We should discuss whether it's something that we want to use in this repo, or whether we should continue to use the current approach for our system-level tests.

**Comment by gravitystorm** [comment]
I've been aware of cucumber for years, and I deliberately avoid using it for any project that I'm involved in. I don't think we should use it here either.

I think a lot of my feelings are also described in these two posts:

* https://www.jimmycuadra.com/posts/please-don-t-use-cucumber/
* https://www.codewithjason.com/recommend-against-cucumber/

Basically, I feel like cucumber is a solution looking for a problem.

First, the natural language syntax isn't actually natural language in any way. It's just a thin textual layer above a bunch of hand-written regular expressions to match the sentences and run some actual test code. So if you make a small change to the sentence, you need to make the corresponding change to the regexp. If you try to make the regexps flexible (e.g. handling both "I am on the page" and "I am looking at the page"), they quickly get overcomplicated. Every new sentence you make needs a corresponding regular expression and implementation. So you end up writing just the same capybara-based tests (like `visit users_path`), but then you also have to wrap them in a regexp-based abstraction layer:

```
Given("I am on the microcosms page") do
  visit microcosms_path
end

Given("I am on the microcosm {string} page") do |name|
  visit microcosm_path(Microcosm.find_by(:name => name))
end
```

Secondly, we don't have any "Product Owners" or "Business Analysts" or anyone in a similar position to that. Even if we did, I doubt that they would want to read or contribute to the tests (something that cucumber claims to enable, but in the real world I think rarely actually happens). They would be writing text into github issue comments and developers would then be writing the relevant tests. So it feels like the layer of abstraction - which I still believe isn't genuinely useful anyway - isn't even necessary in our situation. 

Third, it's just a whole bunch of extra work when creating or reviewing our tests. Any time I see something like `Given there is a microcosm "MappingDC", "Washington, DC, USA", "38.9", "-77.03", "38.516", "39.472", "-77.671", "-76.349"` then I need to find the definition of that sentence, figure out what bits get parsed out into which variables, read the underlying definition to find out what it's really doing (and hope there isn't another definition with a similar regular expression that I should be reading instead), and so on. So I believe it increases the work needed to write tests and brings no benefits, and we'd be better just continuing to write code-based tests.

So I think we should continue to avoid introducing cucumber tests. I'm interested to hear what other people think.

**Comment by tomhughes** [comment]
I've always been somewhat puzzled by those sort of tests - you see them a lot of conference slides and things and I'd always wondered how on earth they actually worked.

Having now read some of the resources you pointed at I now understand - you have to write a whole pile of hidden matchers to translate that DSL into another DSL which is the real test language.

So I think I agree with you - the whole thing seems borderline insane to me. You wind up having to write more code because you have to write the "english" language test and then the translation layer to translate that to capybara or whatever and then you wind up with something that is, to my eyes at least, harder to understand and definitely harder to debug or enhance because of the need to work on both layers at once.

**Comment by lonvia** [comment]
Just to add a data point: I've used BDD-style extensively for [testing Nominatim](https://github.com/openstreetmap/Nominatim/blob/master/test/bdd/db/import/addressing.feature) (after getting the idea from OSRM) and in this context I'm actually quite happy with it. That is because the test all follow a very narrowly defined pattern: fill table A with data -> run import script -> expect data in table B. Nominatim has hundreds of these functional tests and BDD provides a concise notation for them.

Following that success I tried to do other tests in BDD format as well and quickly gave up on it again. Programmers are simply quicker in reading code than reading some English language translation, so BDD gets in the way when writing and reading the tests. I kept forgetting what English language terms I had defined and needed to look them up all the time. And you loose the flexibility of a function that can be called with many different parameters. Everything needs an extra instruction.

Long story short: BDD has a use if you can represent your test suite with only a few instruction patterns where each instruction still represents a fairly complex step and/or where representing your test data in a table is helpful. For everything else: stay away from it.

**Comment by mmd-osm** [comment]
I think @openbrian was very open to the idea of switching from cucumber to something else:

**Comment by mmd-osm** [comment]
I think @openbrian was very open to the idea of switching from cucumber to something else: quoting from https://github.com/openstreetmap/openstreetmap-website/pull/2390#issuecomment-580997440

_I'm going to continue using cucumber as that's how I get from the design document Design.md to test scenarios, to specs, to code. I'm not opposed to minitest. In fact I'm a strong supporter of the testing pyramid, which places unit tests as the most popular testing in your suite. i just need to know where to get started._

He was explicitly asking for help and support to move away from cucumber.

**Comment by gravitystorm** [comment]
I think we've come to the conclusion here to avoid cucumber for the foreseeable future.
